source/tutorials/nixos-configuration-testing.md

```diff
@@ -0,0 +1,485 @@
+It takes one attribute set as its argument.
+The attribute set has two attributes `nodes` and `testScript`.
+The attribute nodes contains a set of named configurations.
+This is the case because test scripts can involve more than one virtual machine.
+
+### minimal example
+
+A minimal test that can be run on the default configuration is to check if the user `root` and `alice` have access to firefox.
+
+Because in our example we only use one virtual machine we simply call it `machine` and place it inside the nodes attribute set.
+
+```nix
+let
+  nixpkgs = <nixpkgs>;
+  pkgs = import nixpkgs {};
+in
+  pkgs.nixosTest {
+    nodes.machine = { config, pkgs, ... }: {
+      ...
+    }
+  }
+```
+<!-- does it make sense to have a un-compilable section? -->
+
+The test script looks like this:
+
+```nix
+machine.wait_for_unit("default.target")
+machine.succeed("su -- alice -c 'which firefox'")
+machine.fail("su -- root -c 'which firefox'")
+```
+
+
+The python script is referring to the `machine` attribute we used to name the configuration in the `nodes` attribute.
+The script waits until start up is reaching systemd `default.target`.
+We are using the `su` command to switch between users and the `which` command to see if the user has access to `firefox`.
+We expect that the command `which firefox` to succeed for user `alice` and to fail for `root`.
+This script is put into a function inside the `testScript` attribute.
+Unlike the virtual machine before we don't need to specify a password, the test script has super user rights in the virtual machine.
+
+The complete `minimaltest.nix` file content looks like the following:
+```{code-block}
+let
+  nixpkgs = <nixpkgs>;
+  pkgs = import nixpkgs {};
+in
+  pkgs.nixosTest {
+    nodes.machine = { config, pkgs, ... }: {
+      boot.loader.systemd-boot.enable = true;
+      boot.loader.efi.canTouchEfiVariables = true;
+
+      services.xserver.enable = true;
+      services.xserver.displayManager.gdm.enable = true;
+      services.xserver.desktopManager.gnome.enable = true;
+
+      users.users.alice = {
+        isNormalUser = true;
+        extraGroups = [ "wheel" ];
+        packages = with pkgs; [
+          firefox
+          thunderbird
+        ];
+      };
+
+      system.stateVersion = "22.05";
+    };
+    testScript = {nodes, ...}: ''
+      machine.wait_for_unit("default.target")
+      machine.succeed("su -- alice -c 'which firefox'")
+      machine.fail("su -- root -c 'which firefox'")
+    '';
+  }
+```
+
+### Running tests
+
+To set up all machines and execute the test script:[^nixdev]
+
+```shell-session
+$ nix-build minimaltest.nix
+```
+
+You'll notice an error message if something goes wrong.[^nixdev]
+
+In case the tests succeed, you should see at the end:[^nixdev]
+
+```shell-session
+...
+test script finished in 10.96s
+cleaning up
+killing machine (pid 10)
+(0.00 seconds)
+/nix/store/bx7z3imvxxpwkkza10vb23czhw7873w2-vm-test-run-unnamed
+```
```

Comment:
I would recommend doing [this pattern](https://github.com/NixOS/nix.dev/pull/267/files#diff-c08ceee24137f72990af94b2e6a5eb487ec1ec767437a6082ba06d9d4244b835R94-R103) for shell input/output. It's both "semantically" readable in source and can be [highlighted with CSS to distinguish input and output.](https://github.com/NixOS/nixos-summer/pull/42) @sbruder maybe you could also add that for nix.dev?